lib/HLSL/DxilModule.cpp

```diff
@@ -367,7 +389,65 @@ void DxilModule::CollectShaderFlags(ShaderFlags &Flags) {
           case DXIL::OpCode::Msad:
             hasMSAD = true;
             break;
-          case DXIL::OpCode::BufferLoad:
+          case DXIL::OpCode::BufferLoad: {
+              Value *resHandle = CI->getArgOperand(DXIL::OperandIndex::kBufferStoreHandleOpIdx);
+              CallInst *handleCall = cast<CallInst>(resHandle);
+
+              if (ConstantInt *resClassArg =
+                  dyn_cast<ConstantInt>(handleCall->getArgOperand(
+                      DXIL::OperandIndex::kCreateHandleResClassOpIdx))) {
+                  DXIL::ResourceClass resClass = static_cast<DXIL::ResourceClass>(
+                      resClassArg->getLimitedValue());
+                  if (resClass == DXIL::ResourceClass::UAV) {
+                    if (ConstantInt *resRangeID =
+                            dyn_cast<ConstantInt>(handleCall->getArgOperand(
+                                DXIL::OperandIndex::kCreateHandleResIDOpIdx))) {
+                      DxilResource &res = GetUAV(resRangeID->getLimitedValue());
+                      if (res.GetKind() == DXIL::ResourceKind::TypedBuffer) {
+                        Value *GV = res.GetGlobalSymbol();
+                        llvm::Type *Ty = GV->getType()->getPointerElementType();
+                        // Check if uav load is multi component load.
+                        if (!IsResourceSingleType(Ty)) {
+                          hasMulticomponentUAVLoads = true;
+                        }
+                      }
+                    }
+                  }
+              }
+              else if (PHINode *resClassPhi = dyn_cast<
+                  PHINode>(handleCall->getArgOperand(
+                      DXIL::OperandIndex::kCreateHandleResClassOpIdx))) {
+                  unsigned numOperands = resClassPhi->getNumOperands();
+                  for (unsigned i = 0; i < numOperands; i++) {
+                      if (ConstantInt *resClassArg = dyn_cast<ConstantInt>(
```

Comment:
Could this block starting here be wrapped in a function and re-used for phi and non-phi cases?  There's also a lot of common code between this and the texture case.  Perhaps we can handle all four with one function?